accelerator/cpu_accelerator.py

```diff
@@ -0,0 +1,271 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+# DeepSpeed Team
+
+import torch
+from deepspeed.accelerator.abstract_accelerator import DeepSpeedAccelerator
+import oneccl_bindings_for_pytorch  # noqa: F401
+import psutil
+
+
+# accelerator for Intel CPU
+class CPU_Accelerator(DeepSpeedAccelerator):
+
+    def __init__(self):
+        self._name = 'cpu'
+        self._communication_backend_name = 'ccl'
+        self.max_mem = psutil.Process().memory_info().rss
+
+        def _check_and_mapping_mpich_env():
+            # Workaround when using Intel MPI launcher
+            # Model can check "RANK" and "WORLD_SIZE" for rank and world size
+            import os
+            if "RANK" not in os.environ and "PMI_RANK" in os.environ:
+                os.environ['RANK'] = os.environ.get('PMI_RANK')
+                print("mapping environment variable PMI_RANK to RANK")
+            if "WORLD_SIZE" not in os.environ and "PMI_SIZE" in os.environ:
+                os.environ['WORLD_SIZE'] = os.environ.get('PMI_SIZE')
+                print("mapping environment variable PMI_SIZE to WORLD_SIZE")
+            if "LOCAL_RANK" not in os.environ and "MPI_LOCALRANKID" in os.environ:
+                os.environ['LOCAL_RANK'] = os.environ.get('MPI_LOCALRANKID')
+                print("mapping environment variable MPI_LOCALRANKID to LOCAL_RANK")
+
+        _check_and_mapping_mpich_env()
+
+    def is_synchronized_device(self):
+        return True
+
+    # Device APIs
+    def device_name(self, device_index=None):
+        return 'cpu'
+
+    def device(self, device_index=None):
+        return None
+
+    def set_device(self, device_index):
+        return
+
+    def current_device(self):
+        return 0
+
+    def current_device_name(self):
+        return 'cpu'
+
+    def device_count(self):
+        from deepspeed.utils.numa import get_numa_cores
+        # Count NUMA node for number of cpu accelerators. On machine with HBM
+        # In flat mode, HBM is in separate NUMA node with no cores on this node.
+        # Ignore these NUMA nodes with no cores.
+        numa_core_lists = get_numa_cores()
+        core_count = 0
+        for core_list in numa_core_lists:
+            if len(core_list) > 0:
+                core_count += 1
+        return core_count
```

Comment:
It seems inconsistent that device  count can be > 1, but other device api functions like `current_device()` behave as if there is only one CPU device. 
I am concerned about client logic such as below that relies on some form of unique device identifier.  Can you please explain why this is fine? 

```python
num_devices = get_accelerator().device_count() 
 all_devices = [num_devices]
  for i, dev in enumerate(all_devices)
        if i == get_accelerator().current_device():
           # Do something about my device
        else:
          # Do something about other devices
```

Reply:
Yes, this will be a concern.   CPU as accelerator is different from discrete device that CPU is more 'divisible'.  For example, two sockets 2x56 core system can be used as a single accelerator with 112 cores, or 2 accelerators with 56 cores, or 4 accelerators with 28 cores.

A better logic is as follow:
1. Multinode runner: let device_count() return env LOCAL_SIZE and current_device() return env LOCAL_RANK.
2. Singlenode runner:  also use LOCAL_SIZE and LOCAL_RANK

If LOCAL_SIZE and LOCAL_RANK are not set, this is the case in runner when deepspeed is launched. (https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/launcher/runner.py#L405),  in this case return number of NUMA nodes on host.  In this situation we assume current_device() will not be called because local rank had not been decided yet.

Reply:
Will change to this new logic.   After the change a simple test result will be following:
1. Single node launching, without specify num_gpus, the number of rank is decided by numa nodes:
```
$ deepspeed --bind_cores_to_rank torch_dist.py
...
rank = 0, size = 2 device_count = 2 current_device = 0
rank = 1, size = 2 device_count = 2 current_device = 1
...

```

2. Two nodes launching, number of devices is specified in hostfile, each host has 2 slots
```
$ deepspeed --bind_cores_to_rank --hostfile=hostfile.txt --launcher impi --master_addr 192.168.10.44 torch_dist.py
...
rank = 0, size = 4 device_count = 2 current_device = 0
rank = 1, size = 4 device_count = 2 current_device = 1
rank = 3, size = 4 device_count = 2 current_device = 1
rank = 2, size = 4 device_count = 2 current_device = 0
...
```